### PR TITLE
dir: make sure pool is mounted

### LINKED ISSUE #3938: DIR storage pool issues after update to 2.18
# Required information

 * Distribution: Ubuntu 16.04.3 LTS
 * The output of "lxc info":
```
config: {}
api_extensions:
- storage_zfs_remove_snapshots
- container_host_shutdown_timeout
- container_syscall_filtering
- auth_pki
- container_last_used_at
- etag
- patch
- usb_devices
- https_allowed_credentials
- image_compression_algorithm
- directory_manipulation
- container_cpu_time
- storage_zfs_use_refquota
- storage_lvm_mount_options
- network
- profile_usedby
- container_push
- container_exec_recording
- certificate_update
- container_exec_signal_handling
- gpu_devices
- container_image_properties
- migration_progress
- id_map
- network_firewall_filtering
- network_routes
- storage
- file_delete
- file_append
- network_dhcp_expiry
- storage_lvm_vg_rename
- storage_lvm_thinpool_rename
- network_vlan
- image_create_aliases
- container_stateless_copy
- container_only_migration
- storage_zfs_clone_copy
- unix_device_rename
- storage_lvm_use_thinpool
- storage_rsync_bwlimit
- network_vxlan_interface
- storage_btrfs_mount_options
- entity_description
- image_force_refresh
- storage_lvm_lv_resizing
- id_map_base
- file_symlinks
- container_push_target
- network_vlan_physical
- storage_images_delete
- container_edit_metadata
- container_snapshot_stateful_migration
- storage_driver_ceph
- storage_ceph_user_name
- resource_limits
- storage_volatile_initial_source
- storage_ceph_force_osd_reuse
- storage_block_filesystem_btrfs
api_status: stable
api_version: "1.0"
auth: trusted
public: false
environment:
  addresses: []
  architectures:
  - x86_64
  - i686
  certificate: |
    -----BEGIN CERTIFICATE-----
    MIIFXDCCA0SgAwIBAgIRAIbnttHqC+0WXXxBnxzQIZowDQYJKoZIhvcNAQELBQAw
    LzEcMBoGA1UEChMTbGludXhjb250YWluZXJzLm9yZzEPMA0GA1UEAwwGcm9vdEBz
    MB4XDTE3MDgyMTIyMDUzN1oXDTI3MDgxOTIyMDUzN1owLzEcMBoGA1UEChMTbGlu
    dXhjb250YWluZXJzLm9yZzEPMA0GA1UEAwwGcm9vdEBzMIICIjANBgkqhkiG9w0B
    AQEFAAOCAg8AMIICCgKCAgEAzXq7ZD2ZHUjCzQhmXIqgbCoXHN2V/WPIU1Gid0oc
    u2/pagGK7ELZTCR6ZbXur3Y0JUPODalbqCRmPjOcAurUc7nijNfb/BpShJ8pIxP0
    +2RpL2aIBS6Sc29x0EA6DBVet5+fE8F3ga7mfEJ96KITK98SaP2c/eq0+LphBgQ8
    68omLc8zFtti+mcYvQnfpTq/3gPUUH53dyz9W/zIOJ2ssYKvRoqmqSrZKJTD3Cd/
    XjC9M1QPMZ44zNFzvVJZNNyIZ/mqqJgYQ6uHt3s6Q8qTI18Dd+j2T8ttmxBnWUPL
    w5TfbjeYaRBQRaVx9QJW+hGNw7pvmsfogXVez+wYBfMqvLtJWKAx92L48T0z1jD0
    I59Myh+U8YHUJZHXhBGfDFb0pbPqnXCwkYpWKCc6bA8Z+gHMevKuDwczhFtLTC0L
    66SKgMaTO+1s9FkAI8wVElY2VnR3PgzdUkNs1m4EZP/JXj0CB3ZZeYx+OJp8LCrx
    xXejLfovzy3ggVZ+3XHDoan+Vf1OlnUdIrm4Yvu6EM9IK3vta2F1ZupFT5DGPlKR
    rj5bZxide527CnYBW53f7uywqu9v7cuytg6lb2vqcok18XUMtLd5HQX0EElaJ9HY
    Ecxf+4E2VLV8dEeHZ8uxhuqjknU5cxHWdViN4hjUI6tQlX4wtGgr5+aixyn3vZ7Q
    B8kCAwEAAaNzMHEwDgYDVR0PAQH/BAQDAgWgMBMGA1UdJQQMMAoGCCsGAQUFBwMB
    MAwGA1UdEwEB/wQCMAAwPAYDVR0RBDUwM4IBc4cEsAloaocQKgEE+AFRUW0AAAAA
    AADwDYcECjIA/ocQKgEE+AFRUW0AAAAAUADwDTANBgkqhkiG9w0BAQsFAAOCAgEA
    AKF+s3JYr1pQrBCZYLg3QD/orZu2SoXWRlZ3RDT3zaIQep8yqrcZ5ACXLbhl1ZZ/
    olu1a+vRm6bRWm5za3u2GoSg+58B8NmSIam83yOPd4ngA7fUcmmzsndL8eR0bmXJ
    QYuWn0SfSs/AbHUjatlPf/TFUJ3fcC0f1wxvwN08K7SXDrBC7vDp9gNin9mFr8P1
    EHrBTnaE3VnPgKtT1lISr90e4b93wlBbUTP7a3MAUJgylvAEz0ig9ED89l7AjG5R
    y4797iTYHAAJEtppG+8/SNjwwbxliypFnkgnE6v+U+ohivQOBP8fXrbkuy46rp8E
    imznh1H5xHuCzDk4JO3Bhkxd7HffPlTMwzcAQbbcL67qHtz+YEp5YWJgYf36qirE
    gq92jsUKYa++QT0bEdG9EcAv2aJvDwR938c4Q8NDb7iAEaeOSSZJ1yq2KdkGMPFU
    O+v8aQCK/EKrDSl0wM2uUHc84LxRIGLvCNSElGFz9Rdv+8on1inDu5z5U0KXtQQ6
    u9efONMfdemtfzeP0hCCc+PTkrKSFxRNkHbLepiKj7bZ1YJqmhLUJ86rafzXJnvu
    ZFzU7bwu1R1/2pNdSF1jEwWScM0vj2N9DwqEvc9nmCEFkaGUWHjipXWwiJF5cLdo
    MPizNFJM6BY4R651t76ZdEZDhWoZAyKvuM18ck0/GY0=
    -----END CERTIFICATE-----
  certificate_fingerprint: 11e33e64377a6e9aaa7ef9316553f4f4124573e4de71f453c9f0282787276ab6
  driver: lxc
  driver_version: 2.1.0
  kernel: Linux
  kernel_architecture: x86_64
  kernel_version: 4.4.0-97-generic
  server: lxd
  server_pid: 4864
  server_version: "2.18"
  storage: dir
  storage_version: "1"
```

# Issue description

My system is using the stable PPA and was originally running 2.13 (I believe).
I have a single DIR storage pool on a separate mount point `/storage/lxd/pool`, which is set as `source` on the `default` storage pool, attached to the `default` profile.

After apt-upgrading (to 2.18) and rebooting, I noticed that none of my containers had started automatically. What I found out is that all the symlinks in `/var/lib/lxd/containers` were pointing to `/var/lib/lxd/storage-pools/default` instead of the `default` pool's source (which is where the containers' rootfs is currently stored).
I then manually recreated the symlinks and, with no surprise, each container started happily.

Anyway, the real issue now is that creating a new container still does not use the correct pool's directory, even thought it seems to be correctly configured.
I could not find any clue by `lxc monitor`ing during container creation, no errors nor warnings...

# Information to attach

Pool configuration:

```
super@s:~$ lxc storage list
+---------+-------------+--------+-------------------+---------+
|  NAME   | DESCRIPTION | DRIVER |      SOURCE       | USED BY |
+---------+-------------+--------+-------------------+---------+
| default |             | dir    | /storage/lxd/pool | 6       |
+---------+-------------+--------+-------------------+---------+
super@s:~$ lxc storage show default
config:
  source: /storage/lxd/pool
description: ""
name: default
driver: dir
used_by:
- /1.0/containers/camrec
- /1.0/containers/deluge
- /1.0/containers/openvpn
- /1.0/containers/test
- /1.0/containers/web
- /1.0/profiles/default
```

Profile configuration:

```
super@s:~$ lxc profile list
+---------+---------+
|  NAME   | USED BY |
+---------+---------+
| default | 5       |
+---------+---------+
super@s:~$ lxc profile show default
config: {}
description: Default LXD profile
devices:
  eth0:
    nictype: bridged
    parent: lxdbr0
    type: nic
  root:
    path: /
    pool: default
    type: disk
name: default
used_by:
- /1.0/containers/camrec
- /1.0/containers/deluge
- /1.0/containers/openvpn
- /1.0/containers/test
- /1.0/containers/web
```

After running `lxc init ubuntu: test`:

```
super@s:~$ lxc config show test --expanded
architecture: x86_64
config:
  image.architecture: amd64
  image.description: ubuntu 16.04 LTS amd64 (release) (20170919)
  image.label: release
  image.os: ubuntu
  image.release: xenial
  image.serial: "20170919"
  image.version: "16.04"
  volatile.apply_template: create
  volatile.base_image: 03c2fa6716b5f41684457ca5e1b7316df520715b7fea0378f9306d16fdc646ee
  volatile.eth0.hwaddr: 00:16:3e:95:27:6a
  volatile.eth0.name: eth0
  volatile.idmap.base: "0"
  volatile.idmap.next: '[{"Isuid":true,"Isgid":false,"Hostid":100000,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":100000,"Nsid":0,"Maprange":65536}]'
  volatile.last_state.idmap: '[{"Isuid":true,"Isgid":false,"Hostid":100000,"Nsid":0,"Maprange":65536},{"Isuid":false,"Isgid":true,"Hostid":100000,"Nsid":0,"Maprange":65536}]'
devices:
  eth0:
    nictype: bridged
    parent: lxdbr0
    type: nic
  root:
    path: /
    pool: default
    type: disk
ephemeral: false
profiles:
- default
stateful: false
description: ""
```

My actual pool storage directory:

```
super@s:~$ ls -Al /storage/lxd/pool/containers/
total 16
drwxr-xr-x+ 4 100000 100000 4096 Aug 22 22:46 camrec
drwxr-xr-x+ 4 100000 100000 4096 Sep 19 20:36 deluge
drwxr-xr-x+ 4 100000 100000 4096 Aug 22 17:23 openvpn
drwxr-xr-x+ 4 100000 100000 4096 Sep 10 22:18 web
```

LXD's symlinks (be aware that I manually recreated the symlinks for my original containers, so focus only on the `test` one):

```
super@s:~$ sudo ls -Al /var/lib/lxd/containers/
total 64
lrwxrwxrwx 1 root root    35 Oct 10 22:06 camrec -> /storage/lxd/pool/containers/camrec
lrwxrwxrwx 1 root root    35 Oct 10 22:06 deluge -> /storage/lxd/pool/containers/deluge
-rw-r--r-- 1 root root 60318 Sep 19 22:33 lxc-monitord.log
lrwxrwxrwx 1 root root    36 Oct 10 21:57 openvpn -> /storage/lxd/pool/containers/openvpn
lrwxrwxrwx 1 root root    50 Oct 12 19:05 test -> /var/lib/lxd/storage-pools/default/containers/test
lrwxrwxrwx 1 root root    32 Oct 10 22:06 web -> /storage/lxd/pool/containers/web
```

```
super@s:~$ sudo ls -Al /var/lib/lxd/storage-pools/default/containers/
total 4
drwxr-xr-x+ 4 root root 4096 Oct 12 19:05 test
```

### PR DESCRIPTION
**Pull request by brauner**
Closes #3938.

Signed-off-by: Christian Brauner <christian.brauner@ubuntu.com>

### CODE REVIEW

**Comment by stgraber** [comment]
That feels much more complicated than it should be. Can't we just setup the bind-mount in StoragePoolInit if it's missing and call it done?

**Comment by stgraber** [comment]
So when LXD first loads the pool, it sets up the bind-mount if needed and that's it. We don't bother to unmount it on exit, the only case where we do unmount it is on pool removal.

**Comment by brauner** [comment]
StoragePoolInit() does not do any on-disk initialization so that we can initialize a pool without requiring it to be mountable or anything.

**Comment by stgraber** [comment]
Still feels very much wrong to do the pool mount/umount for every single operation. The pool should get mounted when it's created or when LXD is first started and that's it.

I wouldn't expect my entire /var/lib/lxd/storage-pools/blah to get unmounted once my last container goes offline, that just feels odd for the dir backend.

**Comment by brauner** [comment]
@stgraber, we do mount the pool on create already. Nothing has changed in that regard.
We shouldn't require the pool to be mounted on LXD startup. That has the problem that you won't be able to start LXD if the pool is messed up which in turn prevents users from doing anything to the pool including deleting it. We've seen these problems before.

File: lxd/storage_dir.go [hunk 1]
```diff
@@ -334,12 +344,17 @@ func (s *storageDir) StoragePoolVolumeDelete() error {
 		return fmt.Errorf("no \"source\" property found for the storage pool")
 	}
 
+	_, err := s.StoragePoolUmount()
+	if err != nil {
+		return err
+	}
```
Comment (stgraber) [inline lxd/storage_dir.go]: Hmm, what? Why would you unmount the entire pool when attempting to delete a storage volume?

Comment (brauner) [reply lxd/storage_dir.go]: Oh, I missed that one. I thought I had removed all umounts. Thanks!

File: lxd/storage_dir.go [hunk 2]
```diff
@@ -642,7 +677,8 @@ func (s *storageDir) ContainerCopy(target container, source container, container
 }
 
 func (s *storageDir) ContainerMount(c container) (bool, error) {
-	return true, nil
+
```
Comment (stgraber) [inline lxd/storage_dir.go]: This is upsetting jenkins (blank first line)

**Pull request merged**